### ISSUE TITLE
MP4Box dash writes $RepresentationID$_$Number$$Time$.m4s media url value.

### ISSUE THREAD

**Issue by Murmur** [gpac/gpac#839]
**Test files**
http://akini.mbnet.fi/videos/test-dash1/
http://akini.mbnet.fi/videos/test-dash1/dash/
http://akini.mbnet.fi/videos/test-dash1/dash/manifest.mpd
http://akini.mbnet.fi/videos/test-dash1/dash/manifest_orig.mpd

MP4Box inserts both $Number$ and $Time$ substitutes to the media url mask value. Segment files use a 1..n sequence numbers (1080p_i.mp4, 1080p_1.m4s, 1080p_2.m4s, 1080p_3.m4s ...).

Manifest works fine if I remove $Time$ variable from the value, see manifest.mpd test link above.

```
<SegmentTemplate timescale="1000" media="$RepresentationID$_$Number$$Time$.m4s" startNumber="1" initialization="$RepresentationID$_i.mp4">
    <SegmentTimeline>
     <S t="0" d="12000" r="6"/>
     <S d="6000"/>
    </SegmentTimeline>
   </SegmentTemplate>

   <SegmentTemplate timescale="1000" media="$RepresentationID$_$Number$$Time$.m4s" startNumber="1" initialization="$RepresentationID$_i.mp4">
    <SegmentTimeline>
     <S t="0" d="11968"/>
     <S d="11989"/>
     <S d="11990"/>
     <S d="11989" r="1"/>
     <S d="11990"/>
     <S d="11989"/>
     <S d="6096"/>
    </SegmentTimeline>
```
```
mp4box.exe -version
MP4Box - GPAC version 0.7.2-DEV-rev86-g81601a2f2-ab-suite
GPAC Copyright (c) Telecom ParisTech 2000-2012
GPAC Configuration: --host=x86_64-w64-mingw32 --build=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --prefix=/local64 --disable-shared --enable-static --static-mp4box
Features: GPAC_64_BITS GPAC_DISABLE_3D

mp4box -dash 12000 -frag 12000 -mem-frags -rap -profile dashavc264:live ^
  -profile-ext "urn:hbbtv:dash:profile:isoff-live:2012" ^
  -mpd-title "" -mpd-info-url "" ^
  -bs-switching no -sample-groups-traf -single-traf -subsegs-per-sidx 1 ^
  -segment-name "$RepresentationID$_$Number$$Init=i$" -segment-timeline ^
  -out "manifest_orig.mpd" ^
  "temp-360p.mp4#trackID=1:id=360p" ^
  "temp-720p.mp4#trackID=1:id=720p" ^
  "temp-1080p.mp4#trackID=1:id=1080p" ^
  "temp-128k.mp4#trackID=1:id=128k"
```

**Comment by rbouqueau** [comment]
$Time$ is inserted because of ```-segment-timeline``` . Could you remove it to make a new test?

**Comment by Murmur** [comment]
New test manifest dashed without a -segment-timeline attribute.
http://akini.mbnet.fi/videos/test-dash1/dash/manifest_notimeline.mpd

I need later run this on various devices(hbbtv) to see if file works properly. Interesting part is a timescale+duration values found in this manifest. Audio aligns to be an exact of 12sec segment. However if run with `-segment-timeline` then manifest gets variable audio durations.
**Video** 153600 / 12800 = 12sec
**Audio** 576000 / 48000 = 12sec

```
  <AdaptationSet segmentAlignment="true" maxWidth="1920" maxHeight="1080" maxFrameRate="25" par="16:9" lang="und">
   <SegmentTemplate timescale="12800" media="$RepresentationID$_$Number$.m4s" startNumber="1" duration="153600" initialization="$RepresentationID$_i.mp4"/>
   <Representation id="360p" mimeType="video/mp4" codecs="avc1.4D4028" width="640" height="360" frameRate="25" sar="1:1" startWithSAP="1" bandwidth="501444">
   </Representation>
   <Representation id="720p" mimeType="video/mp4" codecs="avc1.4D4028" width="1280" height="720" frameRate="25" sar="1:1" startWithSAP="1" bandwidth="1917482">
   </Representation>
   <Representation id="1080p" mimeType="video/mp4" codecs="avc1.4D4028" width="1920" height="1080" frameRate="25" sar="1:1" startWithSAP="1" bandwidth="3026190">
   </Representation>
  </AdaptationSet>
  <AdaptationSet segmentAlignment="true" lang="und">
   <SegmentTemplate timescale="48000" media="$RepresentationID$_$Number$.m4s" startNumber="1" duration="576000" initialization="$RepresentationID$_i.mp4"/>
   <Representation id="128k" mimeType="audio/mp4" codecs="mp4a.40.2" audioSamplingRate="48000" startWithSAP="1" bandwidth="129804">
    <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
   </Representation>
  </AdaptationSet>
```

I have used DASH in a way that `media="$RepresentationID$_$Number$.m4s"` number substitute and `SegmentTimeLine/S` xml elements are used together. Maybe it's not how $Number$ filename scheme is meant to be used? We should use either $Time$+SegmentTimeline/S+timestamp_in_filename or  $Number$+number_in_filename?

**Comment by rbouqueau** [comment]
> Audio aligns to be an exact of 12sec segment. However if run with -segment-timeline then manifest gets variable audio durations.

That's because with segment number we use an average (to match with other implementations). With segment-timeline we use the exact duration. Since AAC frames have a duration of 1024/sample_rate, it is not possible express it with the 1000 timescale in your MPD ; and since we don't want to drift, some segments have a duration that is slightly different from the others.

> I have used DASH in a way that media="$RepresentationID$_$Number$.m4s" number substitute and SegmentTimeLine/S xml elements are used together. Maybe it's not how $Number$ filename scheme is meant to be used? We should use either $Time$+SegmentTimeline/S+timestamp_in_filename or $Number$+number_in_filename?

We usually use either $Number$ or $Time$.

> DolbyDMG tools create a dash using $RepresentationID$__$Number$.m4s+SegmentTimeline/S+number_in_filename

Please paste it here if you want us to have a look.

**Comment by Murmur** [comment]
http://akini.mbnet.fi/videos/test-dash1/dashdmg/
http://akini.mbnet.fi/videos/test-dash1/dashdmg/manifest.mpd

See test files and manifest created by DolbyDMG dash tools. Segment templates use $Number$ substitute, filenames are 360p.mp4, 360p_1.m4s, 360p_2.mp4s, ..., 128k.mp4, 128k_1.m4s, 128k_2.m4s, SegmentTemplate/SegmentTimeline elements are written. (Also audio representation is using an evenly aligned segment duration in a timeline).

**Going back to the topic title** most recent mp4box inserts $Time$ variable to the url template even if input parameter did not specify it. Segment files use 1..n sequence numbering as I wanted to happen. I must remove $Time$ substring from the media url template value.

I found an older mp4box(0.5.2-DEV-rev1117-g0e7107e-master) where $Time$ substitute is not implicitly inserted to the media attribute value. **Is this a regression bug?**
Cmd arguments `-segment-name "$RepresentationID$_$Number$$Init=i$" -segment-timeline`
New mp4box `media="$RepresentationID$_$Number$$Time$.m4s"`
Old mp4box `media="$RepresentationID$_$Number$.m4s"`

This is a snippet from DolbyDMG manifest and I have found this works best in a wide array of hbbtv device models. (evenly aligned audio segment duration + SegmentTemplate/SegmentTimeline element + $Number$ filename sequence numbering)

```
<AdaptationSet mimeType="video/mp4" codecs="avc1.4d4028" segmentAlignment="true" startWithSAP="1">
      <SegmentTemplate timescale="25"
        media="$RepresentationID$_$Number$.m4s" initialization="$RepresentationID$.mp4">
        <SegmentTimeline>
          <S t="0" d="300" r="7"/>
        </SegmentTimeline>
      </SegmentTemplate>
      <Representation id="360p" bandwidth="498840" width="640" height="360" frameRate="25"/>
      <Representation id="720p" bandwidth="1914880" width="1280" height="720" frameRate="25"/>
      <Representation id="1080p" bandwidth="3023592" width="1920" height="1080" frameRate="25"/>
    </AdaptationSet>

<AdaptationSet mimeType="audio/mp4" codecs="mp4a.40.2" lang="und" segmentAlignment="true" startWithSAP="1">
      <Accessibility schemeIdUri="urn:tva:metadata:cs:AudioPurposeCS:2007" value="6"/>
      <Role schemeIdUri="urn:mpeg:dash:role:2011" value="main" />
      <SegmentTemplate timescale="48000"
        media="$RepresentationID$_$Number$.m4s" initialization="$RepresentationID$.mp4">
        <SegmentTimeline>
          <S t="0" d="575488" r="7"/>
        </SegmentTimeline>
      </SegmentTemplate>
      <Representation id="128k" bandwidth="128000" audioSamplingRate="48000">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
      </Representation>
    </AdaptationSet>
```

**Comment by rbouqueau** [comment]
> I found an older mp4box(0.5.2-DEV-rev1117-g0e7107e-master) where $Time$ substitute is not implicitly inserted to the media attribute value. Is this a regression bug?

That's our fault: when using ```-segment-timeline``` with 0.5.2 we didn't check for the according ```$Time$```. It's been added in the last version. We definitely miss checks when parsing arguments but we also miss time to implement them!

**Comment by rbouqueau** [comment]
By the way I added this issue to a list (at GPAC Licensing) related to HbbTV 2 support. Which HbbTV version are you testing against?

**Comment by Murmur** [comment]
Ok thanks. I can always remove a leftover $Time$ substitute from the media attribute value after mp4box dashing. Segment filenames are using 1..n numbers so they fine. I have access to various HbbTV1.0, HbbTV1.5 and HbbTV2.x devices from several manufacturers.

**Comment by jeanlf** [comment]
this should have been fixed by #8d2241a, reopen if needed.
